### PR TITLE
[v1.12 backport] Three fixes for deadlocks

### PR DESCRIPTION
**Pull request by bimmlerd**
- Manual backport of #24672 

The main difference to the upstream PR is the use of older atomic primitives in the third commit, as Go v1.18 does not ship atomic boolean and pointers. 

Once this PR is merged, you can update the PR labels via:
```upstream-prs
$ for pr in 24672; do contrib/backporting/set-labels.py $pr done 1.12; done
```

### CODE REVIEW

**Comment by bimmlerd** [comment]
/test-backport-1.12

**Comment by gandro** [comment]
Conformance AKS failed with a single test failure during the connectivity test. Assuming it's a flake. Restarting.

https://github.com/cilium/cilium/actions/runs/4669862801/jobs/8268888777

**Review by gandro — APPROVED** [review]

**Comment by gandro** [comment]
/ci-aks-1.12

**Comment by gandro** [comment]
This is ready to be merged. Since I'm the backporter and I've done the backport myself, and David is out, I'd appreciate a second pair of eyes on this before it gets merged @joestringer @christarazi or @dylandreimerink (as triager)

The main difference to the upstream commits is the usage of older atomic primitives.

**Review by joestringer — APPROVED** [review]

**Pull request merged**
